etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -173,28 +226,162 @@ process_parcel_file <- function(s3_bucket_uri,
+      angle := calculate_angles(as.matrix(.SD)),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
+    ][
+      ,
+      # Some vertex angles are close to 180 or 360 degrees, which are likely to
+      # be artifacts of the data and not true vertices. We don't want to count
+      # these straight angles in our features, BUT we also can't drop all of
+      # them since some polygons are basically just lines. So, we drop any
+      # straight angle vertices from any polygon with >= 3 non-straight angles
+      # (which is the minimum number of vertices needed to make a polygon)
+      straight_angle := (
+        (
+          data.table::between(angle, 0, 1) |
+            data.table::between(angle, 359, 360)
+        ) | data.table::between(angle, 179, 181)
+      )
+    ][
+      ,
+      num_non_straight_angle := sum(!straight_angle, na.rm = TRUE),
+      by = c("L1", "L2", "L3")
+    ][
+      !straight_angle | num_non_straight_angle < 3 | is.na(angle),
+    ][
+      ,
+      `:=`(
+        # Get edge length using Pythagorean theorem and a rolling lag to get
+        # the distance between each pair of points
+        edge_len = sqrt(
+          (X - data.table::shift(X, type = "lag")) ^ 2 +
+            (Y - data.table::shift(Y, type = "lag")) ^ 2
+        ),
+        # Distance between the centroid and each point in the polygon
+        dist_to_centroid = sqrt(
+          (X - mean(X)) ^ 2 +
+            (Y - mean(Y)) ^ 2
+        ),
+        # Recalculate the angle after removing straight angles
+        angle = calculate_angles(as.matrix(.SD))
+      ),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
+    ]
+
+    # Collapse the vertex-level data back to the parcel level
+    spatial_mat_calc_vert <- spatial_mat_calc[
+      ,
+      .(
+        # Number of vertices in the polygon. Minus 1 because the first and last
+        # vertex are always identical (used to close the polygon)
+        shp_parcel_num_vertices = .N - 1,
+        # Tail is used to drop the first row of each group since it's identical
+        # to the last row (again, same reason as above)
+        shp_parcel_interior_angle_sd = sd(tail(angle, -1), na.rm = TRUE),
+        shp_parcel_centroid_dist_ft_sd = sd(tail(dist_to_centroid, -1)),
+        shp_parcel_edge_len_ft_sd = sd(tail(edge_len, -1), na.rm = TRUE)
+      ),
+      by = "L3"
+    ]
+
+    # Calculate the ratio of the parcel area to the area of its minimum bounding
+    # rectangle
+    spatial_df_rec <- spatial_df_merged %>%
+      st_drop_geometry() %>%
+      st_set_geometry("geometry_3435") %>%
+      mutate(parcel_area = st_area(geometry_3435)) %>%
+      st_minimum_rotated_rectangle() %>%
+      mutate(
+        shp_parcel_mrr_area_ratio = units::drop_units(
+          st_area(geometry_3435) / parcel_area
+        ),
+        # Sometimes the ratio here is very slightly less than 1, even though
+        # that should be impossible. Seems to an artifact of some malformed
```

Comment:
**[Question, non-blocking]** Not a huge deal, but the last sentence is a fragment and I'm wondering what you think might be malformed here?

Reply:
It seems to only be the result when calculating the ratio for these super long, skinny triangles, though I'm really not sure why. The minimum rotated rectangle is in red below. It definitely _should_ have an area larger than the triangle, but it is in fact smaller by 0.00657 ft^2. I don't think it's worth debugging further, given the magnitude of the error. Will update my comment to this effect.

![2754c101-1d09-48ad-9c59-4bd5f8bfe59c](https://github.com/ccao-data/data-architecture/assets/31494343/76f6ba3a-dff9-4173-8392-9a9c7a884b93)